gtda/homology/simplicial.py

```diff
@@ -1772,3 +1772,179 @@ def plot(Xt, sample=0, homology_dimensions=None, plotly_params=None):
             Xt[sample], homology_dimensions=homology_dimensions,
             plotly_params=plotly_params
             )
+
+
+# @adapt_fit_transform_docs
+class LowerStarFlagPersistence(BaseEstimator, TransformerMixin, PlotterMixin):
+    """TODO"""
+    _hyperparameters = {
+        "homology_dimensions": {
+            "type": (list, tuple),
+            "of": {"type": int, "in": Interval(0, np.inf, closed="left")}
+            },
+        "coeff": {"type": int, "in": Interval(2, np.inf, closed="left")},
+        "extended": {"type": bool},
+        "infinity_values": {"type": (Real, type(None))},
+        "reduced_homology": {"type": bool},
+        "collapse_edges": {"type": bool}
+        }
+
+    def __init__(self, homology_dimensions=(0, 1), coeff=2, extended=True,
+                 infinity_values=np.inf, reduced_homology=False,
+                 collapse_edges=False, n_jobs=None):
+        self.homology_dimensions = homology_dimensions
+        self.coeff = coeff
+        self.extended = extended
+        self.infinity_values = infinity_values
+        self.reduced_homology = reduced_homology
+        self.collapse_edges = collapse_edges
+        self.n_jobs = n_jobs
+
+    def _lower_star_extended_diagram(self, X):
+        n_points = max(X.shape)
+        n_points_cone = n_points + 1
+        n_diag = min(X.shape)
+        X = X.tocoo().astype(np.float32)  # ripser needs single precision
+
+        data_diag = np.zeros(n_points_cone, dtype=X.dtype)
+        data_diag[:n_diag] = X.diagonal()
+        max_value = data_diag[:n_diag].max()
+        min_value = data_diag[:n_points].min()
+        data_diag[-1] = min_value - 1
+
+        off_diag = X.row != X.col
+        row_off_diag, col_off_diag = X.row[off_diag], X.col[off_diag]
+
+        row = np.concatenate([row_off_diag, np.arange(n_points)])
+        col = np.concatenate([col_off_diag, np.full(n_points, n_points)])
+        data = np.concatenate([
+            np.maximum(data_diag[row_off_diag], data_diag[col_off_diag]),
+            np.sign(max_value)*2 * max_value + 1 - data_diag[:n_points]
```

Comment:
I added the `sign` bit, since here, we assume that `2*x >x`, which is only true for positive integers

Reply:
Well spotted that there is a problem with negative values!  As for the fix though, I wonder if it would be easier to understand if we did something like
```
min_value + 2 * (max_value - min_value) + 1 - data_diag[:n_points]
```
because I think (?) the formula I initially wrote basically assumed the minimum value was zero and the maximum was positive.

Reply:
Sure! Your proposition is better - i haven't though well enough about my proposal :D

Reply:
Actually, i think it's
``` min_value + 2 * (max_value - min_value) + 1 - (data_diag[:n_points]-min_value)```
which is exactly the same as you had in the beginning! Sorry :)

Reply:
Oh right!